openmeter/productcatalog/plan/service/plan.go

```diff
@@ -0,0 +1,295 @@
+		return pagination.PagedResponse[plan.Plan]{}, fmt.Errorf("invalid list Plans params: %w", err)
+	}
+
+	return s.adapter.ListPlans(ctx, params)
+}
+
+func (s service) CreatePlan(ctx context.Context, params plan.CreatePlanInput) (*plan.Plan, error) {
+	if err := params.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid create Plan params: %w", err)
+	}
+
+	logger := s.logger.With(
+		"operation", "create",
+		"namespace", params.Namespace,
+		"plan.key", params.Key,
+	)
+
+	logger.Debug("creating Plan")
+
+	p, err := s.adapter.CreatePlan(ctx, params)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Plan: %w", err)
+	}
+
+	logger.With("plan.id", p.ID).Debug("Plan created")
+
+	return p, nil
+}
+
+func (s service) DeletePlan(ctx context.Context, params plan.DeletePlanInput) error {
+	if err := params.Validate(); err != nil {
+		return fmt.Errorf("invalid delete Plan params: %w", err)
+	}
+
+	logger := s.logger.With(
+		"operation", "delete",
+		"namespace", params.Namespace,
+		"plan.id", params.ID,
+	)
+
+	// TODO(chrisgacsal): add check which makes sure that Plans with active Subscriptions are not deleted.
+
+	logger.Debug("deleting Plan")
+
+	err := s.adapter.DeletePlan(ctx, params)
+	if err != nil {
+		return fmt.Errorf("failed to delete Plan: %w", err)
+	}
+
+	logger.Debug("Plan deleted")
+
+	return nil
+}
+
+func (s service) GetPlan(ctx context.Context, params plan.GetPlanInput) (*plan.Plan, error) {
+	if err := params.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid get Plan params: %w", err)
+	}
+
+	logger := s.logger.With(
+		"operation", "get",
+		"namespace", params.Namespace,
+		"plan.id", params.ID,
+		"plan.key", params.Key,
+		"plan.version", params.Version,
+	)
+
+	logger.Debug("fetching Plan")
+
+	p, err := s.adapter.GetPlan(ctx, params)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Plan: %w", err)
+	}
+
+	logger.Debug("Plan fetched")
+
+	return p, nil
+}
+
+func (s service) UpdatePlan(ctx context.Context, params plan.UpdatePlanInput) (*plan.Plan, error) {
+	if err := params.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid update Plan params: %w", err)
+	}
+
+	logger := s.logger.With(
+		"operation", "update",
+		"namespace", params.Namespace,
+		"plan.id", params.ID,
+	)
+
+	logger.Debug("updating Plan")
+
+	p, err := s.adapter.GetPlan(ctx, plan.GetPlanInput{
+		NamespacedID: models.NamespacedID{
+			Namespace: params.Namespace,
+			ID:        params.ID,
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Plan: %w", err)
+	}
+
+	allowedPlanStatuses := []plan.PlanStatus{plan.DraftStatus, plan.ScheduledStatus}
+	planStatus := p.Status()
+	if lo.Contains(allowedPlanStatuses, p.Status()) {
+		return nil, fmt.Errorf("only Plans in %+v can be updated, but it has %s state", allowedPlanStatuses, planStatus)
+	}
+
+	logger.Debug("updating plan")
+
+	// NOTE(chrisgacsal): we only allow updating the state of the Plan via Publish/Archive
```

Comment:
What do you think of having separate methods for them instead?

Reply:
Hmm. I am not sure what you are referring to. We do have separate mthods for publish, archive.